feat: Adding EMT reduction.

mrucker opened the pull request:
This pull request contains an implementation of the EMT reduction (https://arxiv.org/abs/2210.14077) along with unit tests.

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,928 @@
+float median(std::vector<float>& array)
+{
+  int size = array.size();
+  std::sort(array.begin(), array.end());
```
is it beneficial to have linear complexity algorithm here?
i.e. stl ones: https://cplusplus.com/reference/algorithm/nth_element/

mrucker replied:
Oh, nice, I didn't know about that method. That's an easy change to make to be safe. Jack had also mentioned maybe there was an existing algorithm in VW that is even better? For what it is worth, in practice I don't think it'll matter too much. We only calculate the median when we split a node in our binary tree, which means it's calculated with log(N) frequency. And when we do calculate it is only on the items in the leaf node being split.

ataymano commented on vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h:
```diff
@@ -0,0 +1,123 @@
+
+struct emt_node
+{
+  emt_node* parent; // parent node
```
are we using parent anywhere?

mrucker replied:
Good catch. No I don't think we are. I think I just added that out of habit.

mrucker replied:
I've cleaned this up.

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,935 @@
+
+  if (size % 2 == 0) {
+    std::nth_element(array.begin(), nth  , array.end());
+    std::nth_element(array.begin(), nth-1, array.end());
```
>>The other elements are left without any specific order, except that none of the elements preceding nth are greater than it, and none of the elements following it are less.

Looks like this call is invalidating guaraantees about nth from previous line.
We probably should either:
1) get the *nth value between these calls
2) replace second nth_element call with something like (constant factor seems to be better here)
```
auto n_minus_1th = std::max_element(array.begin(), array.begin() + size / 2 - 1)
```

mrucker replied:
I grabbed the nth between the calls. I wasn't entirely sure grabbing the max was guaranteed to work given nth_element. Feel free to change it.

mrucker commented:
I finished a major cleaning pass. I think EMT is nearly done. With this pass I:

1. Removed all new statements and replaced then with VW::make_unique
2. Removed all class attributes which weren't actually used (e.g., parent)
3. Refactored to remove two different enums that weren't actually needed
4. Removed unneeded destructors now that we're using make_unique
5. Replaced hacky uses of VW internal classes with std data types
6. Added a lot of unit tests
7. Replaced all couts with trace_message
8. Validated the performance of this implementation against the 206 datasets used in the original publication

ataymano commented on vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h:
```diff
@@ -0,0 +1,111 @@
+
+emt_feats emt_normalize(emt_feats);
+
+emt_feats emt_scale_add(float const, emt_feats const, float const, emt_feats const);
```
Looks like we are doing copying of emt_feats objects in all functions where emt_feats is part of signature.
Can it be changed to references or pointers?

mrucker replied:
Yeah, that's a really good idea. Almost certainly. I'll make a performance pass over the code and improve this.

mrucker replied:
I just finished an optimization pass to tune all function signatures (along with a few other optimizations). 

On a large test data set EMT now runs in about 4 seconds where it used to run in about 6 seconds.

jackgerrits commented on vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h:
```diff
@@ -49,6 +49,24 @@ emt_feats emt_scale_add(float, const emt_feats&, float, const emt_feats&);
 
 emt_feats emt_router_eigen(std::vector<emt_feats>&, VW::rand_state&);
 
+template<typename I, typename R>
```
Nit: `I` is too short. At least `It` or better, `RandomIt` is preferred since ideally we want a random access iterator here

mrucker replied:
Good suggestion. Made the recommended change.

jackgerrits commented on vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h:
```diff
@@ -49,6 +49,24 @@ emt_feats emt_scale_add(float, const emt_feats&, float, const emt_feats&);
 emt_feats emt_router_eigen(std::vector<emt_feats>&, VW::rand_state&);
 
+template<typename I, typename R>
+void emt_shuffle(I first, I last, R* rng)
```
Based on the implementation this will only work with a `VW::rand_state`, so I think we should just remove the template for `R` and use that type.

mrucker replied:
Agreed, that's what I had originally but the compiler was yelling at me.

> invalid use of incomplete type ‘class VW::rand_state’

Any suggestions on this?

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+
+emt_tree::~emt_tree()
+{
+  delete ex->interactions;
```
can nullptr dereferencing happen here in case of 
````
{
  emt_tree a;
}
```

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+  begin = 0;
+
+  all = nullptr;
+  ex = nullptr;
```
can be removed?

mrucker replied:
Or you mean begin? It looks like it.

mrucker replied:
If you see any small refactors like this feel free to make changes yourself.

ataymano replied:
I am sorry for short comment. I was meaning "ex=nullptr" which is initialized 2 lines later.

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+
+  if (!options.add_parse_and_check_necessary(new_options)) { return nullptr; }
+
+  t->scorer_type = static_cast<VW::reductions::eigen_memory_tree::emt_scorer_type>(scorer_type);
```
it looks a bit dangerous (i e if someone adds new scorer_type in the middle). Probably helper conversion from string constant to enum (either with switch/case or using std::map) with explicit error for unknown names is a bit better here

mrucker replied:
@ataymano That's why I explicitly set the values of the enum... Does it still seem dangerous to you?

jackgerrits commented on vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h:
```diff
@@ -0,0 +1,127 @@
+emt_feats emt_router_eigen(std::vector<emt_feats>&, VW::rand_state&);
+
+template <typename RandomIt>
+void emt_shuffle(RandomIt first, RandomIt last, VW::rand_state* rng)
```
We expect this to be non-null and we don't check for null, so we should use a reference instead.

jackgerrits commented on vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h:
```diff
@@ -0,0 +1,127 @@
+
+namespace eigen_memory_tree
+{
+using emt_feats = std::vector<std::pair<int, float>>;
```
Are these hashes? If so, then we should use the same type as elsewhere `uint64_t`

jackgerrits commented on vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h:
```diff
@@ -0,0 +1,127 @@
+  std::list<K> list;
+  std::unordered_map<K, V> map;
+
+  unsigned long max_size;
```
Explicitly sized types should be used instead. e.g. `uint64_t`

jackgerrits commented on vowpalwabbit/core/include/vw/core/reductions/eigen_memory_tree.h:
```diff
@@ -0,0 +1,127 @@
+
+  std::unique_ptr<VW::example> ex;  // we create one of these which we re-use so we don't have to reallocate examples
+
+  long begin;  // for timing performance
```
Can this be an explicitly sized type?

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+{
+////////////////////////////definitions/////////////////////
+////////////////////////////////////////////////////////////
+emt_example::emt_example() { label = 0; }
```
IMO, just set label to 0 in the class definition and then don't define a constructor and let the compiler do it

mrucker replied:
@jackgerrits I don't think I can do this because there is another constructor defined on this class.

jackgerrits replied:
In that case you can just mark it in the header like so:
```c++
struct emt_example
{
  emt_feats base;  // base example only includes the base features without interaction flags
  emt_feats full;  // full example includes the interactions that were passed in as flags
  uint32_t label = 0;

  emt_example() = default // <-----------
  emt_example(VW::workspace&, VW::example*);
};

```

mrucker replied:
ooh fancy, I like it

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+
+emt_node::emt_node()
+{
+  left = nullptr;
+  right = nullptr;
+  router_decision = 0;
```
Same comment here about doing it in the class def

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+
+///////////////////////Helper/////////////////////////////
+//////////////////////////////////////////////////////////
+float emt_median(std::vector<float>& array)
```
This can be a const input

mrucker replied:
@jackgerrits Can it? My compiler complains. I think it is because we re-order the array?

jackgerrits replied:
You're right, I didn't realize `std::nth_element` mutates the input.

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+float emt_median(std::vector<float>& array)
+{
+  int size = array.size();
+  auto nth = array.begin() + size / 2;
```
```suggestion
  const auto nth = array.begin() + size / 2;
```

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+    std::nth_element(array.begin(), nth, array.end());
+    auto v1 = (*nth);
+    std::nth_element(array.begin(), nth - 1, array.end());
+    auto v2 = *(nth - 1);
```
```suggestion
    const auto v2 = *(nth - 1);
```

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+float emt_inner(const emt_feats& xs, const emt_feats& ys)
+{
+  float sum = 0;
+  unsigned long xi = 0;
+  unsigned long yi = 0;
```
Prefer explicit sized types, unless it needs to be `size_t` which looks like it may be the case.

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+////////////////////////////////////////////////////////////////////
+void save_load_examples(emt_node& n, io_buf& model_file, bool& read, bool& text, std::stringstream& msg)
+{
+  WRITEITVAR(n.examples.size(), "n_examples", n_examples);
```
`WRITEITVAR` should not be used. Please use utility functions in `VW::model_utils`

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+  {
+    for (auto& p : e->base)
+    {
+      WRITEIT(p.first, "base_index");
```
`WRITEIT` should not be used. Please use utility functions in `VW::model_utils`

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+{
+  for (auto& p : n.router_weights)
+  {
+    WRITEIT(p.first, "router_index");
```
`WRITEIT` should not be used. Please use utility functions in `VW::model_utils`

jackgerrits commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+std::unique_ptr<emt_node> save_load_node(
+    emt_tree& b, std::unique_ptr<emt_node> n, io_buf& model_file, bool& read, bool& text, std::stringstream& msg)
+{
+  WRITEITVAR(!read && !n, "is_null", is_null);
```
Essentially, you need to implement `read_model_field` and `write_model_field` for your new types to allow for serialization/deserialization.

jackgerrits replied:
https://github.com/VowpalWabbit/vowpal_wabbit/blob/5ca2a7f54a9bf02794c1417b68439538d638555c/vowpalwabbit/core/src/reductions/epsilon_decay.cc#L274

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+    // when we receive ex1 and ex2 their features are indexed on top of eachother. In order
+    // to make sure VW recognizes the features from the two examples as separate features
+    // we apply a map of multiplying by 2 and then offseting by 1 on the second example.
+    for (auto& j : out.feature_space['x'].indices) { j = j * 2; }
```
is it safe to get above (potentially) the maximum hash here?

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+  if (b.scorer_type == emt_scorer_type::self_consistent_rank)
+  {
+    out.indices.clear();
+    out.indices.push_back('x');
```
Can you please define constants for 'x'/'z'?

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+  }
+}
+
+void scorer_example(emt_tree& b, const emt_example& ex1, const emt_example& ex2)
```
probably outside of the scope of this pr, but to consider in the future: as far as i understand randomness is the only reason for emt_tree being nonconst here, so maybe it is worth to consider having rng as mutable.

mrucker replied:
@ataymano Yes, I wouldn't have the time for that.

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+  }
+}
+
+void scorer_example(emt_tree& b, const emt_example& ex1, const emt_example& ex2)
```
just to confirm what kind of asymmetry i mean:
scorer_example({0: a, 1: b}, {0: c})  = some_f({0: a, 1: c, 2: b})
scorer_example({0:c}, {0: a, 1: b}) = some_f({0: c, 1: a, 3: b})

mrucker replied:
Good catch. The asymmetry is not desired.  Here's what we really wanted...

scorer_example({0: a, 1: b}, {0: c}) = some_f({00:ac, 10:bc})

It that possible using vw's interactions. I think I did the offset because it seemed to me like VW wasn't calculating the interactions at all for features in different namespaces with identical indexes (which I know is a strange use case to begin with).

That is, users could say --interactions

mrucker replied:
(also worth noting, before moving all this to VW we were manually calculating these features so this wasn't an issue)

mrucker commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,787 @@
+               .set_learn_returns_prediction(true)  // we set this to true otherwise bounding doesn't work as well
+               .set_end_pass(end_pass)
+               .set_save_load(save_load)
+               .set_output_prediction_type(pred_type)
+               .set_input_label_type(label_type);
```
@jackgerrits What would the input prediction type be? Is that like an initial prediction that we would update?

jackgerrits replied:
When you call `base.predict` what is the prediction type you are expecting from below this reduction? Similarly, when calling `base.learn` what are you setting the label to for that call?

mrucker replied:
got it: when I call base.predict and base.learn I expect/provide a simple regression label.

jackgerrits commented on vowpalwabbit/core/tests/eigen_memory_tree_test.cc:
```diff
@@ -328,4 +328,34 @@ TEST(emt_tests, test_emt_shuffle)
+  {
+    auto* ex = VW::read_example(*vw_save, std::to_string(i) + " | " + std::to_string(i));
+    vw_save->learn(*ex);
+    VW::finish_example(*vw_save, *ex);
```
You need to call `vw_save->finish_example(*ex)` instead of this.

ataymano commented:
was playing around to understand better and the following input data is crashing release build (but not crashing debug one):
```
n = 10
m = 10
with open('data.txt', 'w') as f:
    for i in range(n):
        for j in range(m):
            v = j / m
            l = int(v * 2)
            f.write(f'{l} | 1:{v} 5:{v}\n')
```
Looks like there is ub somewhere

ataymano commented:
playing with it to understand better.
On the following data:
```
n = 10
m = 10
with open('data.txt', 'w') as f:
    for i in range(n):
        for j in range(m):
            v = j / m
            l = int(v * 2)
            f.write(f'{l} | 2:{v} 100:{v}\n')
```
following command line
```
-d data.txt --readable_model r_model.txt --emt --noconstant
```
is crashing with 
![image](https://user-images.githubusercontent.com/41013086/207894263-e979336a-b5eb-4617-9845-f2b35dda00a2.png)
looking.

ataymano commented on vowpalwabbit/core/src/reductions/eigen_memory_tree.cc:
```diff
@@ -0,0 +1,815 @@
+size_t write_model_field(
+    io_buf& io, const reductions::eigen_memory_tree::emt_node& node, const std::string& upstream_name, bool text)
+{
+  size_t bytes = 0;
```
readable model looks like that now:
```
...
emtrootrightrightrightright.is_null() = false
emtrootrightrightrightrightrouter_decision = 0.3114859461784363
emtrootrightrightrightrightleft.is_null() = false
emtrootrightrightrightrightleftrouter_decision = 0
emtrootrightrightrightrightleftleft.is_null() = true
emtrootrightrightrightrightleftright.is_null() = true
emtrootrightrightrightrightleftrouter_weights.size() = 0
emtrootrightrightrightrightleftexamples.size() = 4
emtrootrightrightrightrightleftexamples[0].is_null() = false
emtrootrightrightrightrightleftexamples[0]base.size() = 3
emtrootrightrightrightrightleftexamples[0]base[0].first = 2
emtrootrightrightrightrightleftexamples[0]base[0].second = 0.07031249
emtrootrightrightrightrightleftexamples[0]base[1].first = 100
emtrootrightrightrightrightleftexamples[0]base[1].second = 0.07031249
emtrootrightrightrightrightleftexamples[0]base[2].first = 116060
emtrootrightrightrightrightleftexamples[0]base[2].second = 1
emtrootrightrightrightrightleftexamples[0]full.size() = 3
emtrootrightrightrightrightleftexamples[0]full[0].first = 2
emtrootrightrightrightrightleftexamples[0]full[0].second = 0.07031249
emtrootrightrightrightrightleftexamples[0]full[1].first = 100
emtrootrightrightrightrightleftexamples[0]full[1].second = 0.07031249
emtrootrightrightrightrightleftexamples[0]full[2].first = 116060
emtrootrightrightrightrightleftexamples[0]full[2].second = 1
emtrootrightrightrightrightleftexamples[0]label = 1
emtrootrightrightrightrightleftexamples[1].is_null() = false
emtrootrightrightrightrightleftexamples[1]base.size() = 3
emtrootrightrightrightrightleftexamples[1]base[0].first = 2
emtrootrightrightrightrightleftexamples[1]base[0].second = 0.062499993
emtrootrightrightrightrightleftexamples[1]base[1].first = 100
emtrootrightrightrightrightleftexamples[1]base[1].second = 0.062499993
emtrootrightrightrightrightleftexamples[1]base[2].first = 116060
emtrootrightrightrightrightleftexamples[1]base[2].second = 1
emtrootrightrightrightrightleftexamples[1]full.size() = 3
emtrootrightrightrightrightleftexamples[1]full[0].first = 2
...
```
should we separate upstream from field name with '.'?

jackgerrits replied:
Yeah I think that would help readability a lot here

ataymano commented:
> 

line 55 from screenshot is trying to dereference .end() iterator in case of empty example (i.e. "0 | x:0" with "--noconstant"). pushing fix.
